source/common/tcp/async_tcp_client_impl.cc

```diff
@@ -109,28 +108,33 @@ void AsyncTcpClientImpl::reportConnectionDestroy(Network::ConnectionEvent event)
 void AsyncTcpClientImpl::onEvent(Network::ConnectionEvent event) {
   if (event == Network::ConnectionEvent::RemoteClose ||
       event == Network::ConnectionEvent::LocalClose) {
-    if (disconnected_) {
+    if (connected_) {
+      cluster_info_->trafficStats()->upstream_cx_active_.dec();
+    } else {
       cluster_info_->trafficStats()->upstream_cx_connect_fail_.inc();
     }
 
-    if (!disconnected_ && conn_length_ms_ != nullptr) {
+    if (connected_ && conn_length_ms_ != nullptr) {
       conn_length_ms_->complete();
       conn_length_ms_.reset();
     }
+
     disableConnectTimeout();
     reportConnectionDestroy(event);
-    disconnected_ = true;
+
+    connected_ = false;
     if (connection_) {
       detected_close_ = connection_->detectedCloseType();
     }
 
     dispatcher_.deferredDelete(std::move(connection_));
     if (callbacks_) {
       callbacks_->onEvent(event);
-      callbacks_ = nullptr;
     }
   } else {
-    disconnected_ = false;
+    cluster_info_->trafficStats()->upstream_cx_active_.inc();
```

Comment:
we may need to increase this upstream_cx_active_ during `connect()` to make the meaning same as [here](https://github.com/envoyproxy/envoy/blob/bc0190146f08458ac5a58af5b3545b24660fa8f8/source/common/conn_pool/conn_pool_base.cc#L784)

Reply:
makes sense, thanks